### ISSUE TITLE
User property from future set on View Loads

### ISSUE THREAD

**Issue by geranzo** [DataDog/dd-sdk-android#1836]
**Problem description**
We have a custom user property that we update with the list of screens in navigation back stack. This is done every time the current screen changes, whether navigating forward or backward, just before we call `RumMonitor::startView(...)` to report a new `View Load`. This is working fine for our custom actions, the user property seen in their attributes is being updated correctly. However we noticed this is not the case for the View Loads. Check out below the part of session, the View Load of `Home` screen at 4.36s has already a `user_nav_back_stack` user property updated with the data that was set over a minute later, at 1 min 21 s when the `Profile` View Load happened.

![image](https://github.com/DataDog/dd-sdk-android/assets/4363310/36672c98-d270-4ad0-b730-29b8ccd50241)

**Describe what you expected:**
We'd expect custom user properties to be up to date for View Loads in the same manner as they are for the custom actions.

**Additional context**
 - Datadog SDK version: v2.5.0
 - Datadog Gradle version: v1.12.0

**Comment by 0xnm** [comment]
Hello @geranzo!

Can you please share the way you set these user properties?

Is it done like that:

```
val propertiesA = mapOf("a" to "b")
GlobalRumMonitor.get().addAction(..., propertiesA)
val propertiesB = mapOf("c" to "d")
GlobalRumMonitor.get().startView(..., propertiesB)
```
or you are utilizing `Datadog.setUserInfo` to set them?

Basically code snippets will be useful to understand the setup you have. By looking on the description, it seems that you also start/stop views manually, not using our built-in tracking strategies?

**Comment by geranzo** [comment]
> Can you please share the way you set these user properties?

This is our method to set a user property:

```
    override fun setUserProperty(name: String, value: Any?) {
        if (value != null) {
            globalRum.get().addAttribute(name, value)
        } else {
            removeUserProperty(name)
        }
    }
```

> Basically code snippets will be useful to understand the setup you have. By looking on the description, it seems that you also start/stop views manually, not using our built-in tracking strategies?

Yes, we call `globalRum.get().startView(...)` ourselves on a callback from Compose Navigation library.

**Comment by 0xnm** [comment]
Thanks for a quick response. Can you please expand `removeUserProperty` method? Does it call `globalRum.get().removeAttribute`? And at which point do you call `setUserProperty` - is it only navigation events?

Also is it possible that calls to send custom action are coming from non-UI thread (shouldn't be an issue, but just for the understanding)?

> Yes, we call globalRum.get().startView(...) ourselves on a callback from Compose Navigation library.

We have `dd-sdk-android-compose` with `NavigationViewTrackingEffect` to auto-instrument Compose Navigation. If it doesn't work for your setup, we are open to improve it per your feedback.

**Comment by geranzo** [comment]
> Can you please expand `removeUserProperty` method? Does it call `globalRum.get().removeAttribute`?

Yes, it does:
```
    override fun removeUserProperty(name: String) {
        globalRum.get().removeAttribute(name)
    }
```

> And at which point do you call `setUserProperty` - is it only navigation events?

No, we have a couple more properties that we want to track for the user like e.g. state of a ringer mode or DnD state, theme light or dark etc.

> Also is it possible that calls to send custom action are coming from non-UI thread (shouldn't be an issue, but just for the understanding)?

No, every time we are about to call some API like `addAttribute`, `startView` or `addAction` etc., we launch a coroutine on `Dispatchers.Main.immediate`.

> We have `dd-sdk-android-compose` module with `NavigationViewTrackingEffect` to auto-instrument Compose Navigation. If it doesn't work for your setup, we are open to improve it per your feedback.

It wasn't probably available at the time we were integrating with DD.

**Comment by 0xnm** [comment]
Thanks for details! The problem is clear and we will try to address it.

**Comment by 0xnm** [comment]
The fix is a part of the [2.6.0 release](https://github.com/DataDog/dd-sdk-android/releases/tag/2.6.0).

**Comment by geranzo** [comment]
Thank you! Very much appreciated.
